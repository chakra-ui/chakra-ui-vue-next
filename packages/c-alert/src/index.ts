import { defineComponent, computed, createVNode, mergeProps, isVNode, Fragment } from 'vue';
import { createStylesContext, useMultiStyleConfig, chakra } from '@chakra-ui/vue-system';
import { createContext, getValidChildren } from '@chakra-ui/vue-utils';
import { createIconComponent } from '@chakra-ui/c-icon';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var CInfoIcon = createIconComponent("info");
var CCheckIcon = createIconComponent("check-circle");
var CWarningIcon = createIconComponent("warning-alt");
var CErrorIcon = createIconComponent("warning");

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var STATUSES = {
  info: {
    colorScheme: "blue",
    icon: CInfoIcon
  },
  success: {
    colorScheme: "green",
    icon: CCheckIcon
  },
  warning: {
    colorScheme: "orange",
    icon: CWarningIcon
  },
  error: {
    colorScheme: "red",
    icon: CErrorIcon
  },
  loading: {
    icon: CInfoIcon,
    colorScheme: "blue"
  }
};

var _createStylesContext = createStylesContext("Alert"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = createContext({
  name: "AlertContext",
  errorMessage: "useAlertContext: `context` is undefined. Seems you forgot to wrap alert components in `<c-alert />`"
}),
    AlertProvider = _createContext[0],
    useAlertContext = _createContext[1];
/**
 * CAlert component
 *
 * This is the container component for all Alert components.
 * It also provides state and context to it's compound components
 */


var CAlert = defineComponent({
  name: "CAlert",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    status: {
      type: [String],
      "default": "info"
    },
    colorScheme: {
      type: [String]
    },
    styleConfig: {
      type: [Object]
    },
    variant: {
      type: [String],
      "default": "solid"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var colorScheme = computed(function () {
      var _STATUSES$props$statu;

      return props.colorScheme || ((_STATUSES$props$statu = STATUSES[props == null ? void 0 : props.status]) == null ? void 0 : _STATUSES$props$statu.colorScheme);
    });
    var themingProps = computed(function () {
      return {
        colorScheme: colorScheme.value,
        variant: props.variant
      };
    });
    AlertProvider({
      status: computed(function () {
        return props.status;
      })
    });
    var styles = useMultiStyleConfig("Alert", themingProps);
    StylesProvider(styles);
    var alertStyles = computed(function () {
      return _extends({
        width: "100%",
        display: "flex",
        alignItems: "center",
        position: "relative",
        overflow: "hidden"
      }, styles.value.container);
    });
    return function () {
      return createVNode(chakra.div, mergeProps({
        "role": "alert",
        "__label": "alert",
        "__css": alertStyles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertTitle component
 *
 * The title component for alerts
 */

var CAlertTitle = defineComponent({
  name: "CAlertTitle",
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__title",
        "__css": styles.value.title
      }, attrs), _isSlot(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
/**Â«
 * CAlertDescription component
 *
 * The description component for alerts
 */

var CAlertDescription = defineComponent({
  name: "CAlertDescription",
  setup: function setup(_, _ref3) {
    var attrs = _ref3.attrs,
        slots = _ref3.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__description",
        "__css": styles.value.description
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertIcon component
 *
 * The Icon component for alerts
 */

var CAlertIcon = defineComponent({
  name: "CAlertIcon",
  setup: function setup(_, _ref4) {
    var attrs = _ref4.attrs,
        slots = _ref4.slots;
    var styles = useStyles();

    var _useAlertContext = useAlertContext(),
        status = _useAlertContext.status;

    var BaseIcon = STATUSES[status.value].icon;
    var css = computed(function () {
      return status.value === "loading" ? styles.value.spinner : styles.value.icon;
    });
    return function () {
      var validChildren = getValidChildren(slots);
      return createVNode(chakra.span, mergeProps({
        "display": "inherit",
        "__label": "alert__icon"
      }, attrs, {
        "__css": css.value
      }), {
        "default": function _default() {
          return createVNode(Fragment, null, [validChildren.length ? slots : createVNode(BaseIcon, {
            "h": "100%",
            "w": "100%"
          }, null)]);
        }
      }); // return <icon {...styles.value.icon} {...attrs}></icon>
    };
  }
});

export { CAlert, CAlertDescription, CAlertIcon, CAlertTitle };
