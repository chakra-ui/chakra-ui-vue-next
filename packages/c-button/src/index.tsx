import { defineComponent } from "vue"

export { default as CButton } from "./button"
export { default as CButtonGroup } from "./button-group"
export { default as CIconButton } from "./icon-button"
