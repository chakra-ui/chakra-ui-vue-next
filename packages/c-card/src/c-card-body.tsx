import { defineComponent, h } from "vue"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"
import { useCardStyles } from "./card.context"

export interface CCardBodyProps extends HTMLChakraProps<"div"> {}

export const CCardBody = defineComponent({
  name: "CCardBody",
  setup(_, { attrs, slots }) {
    const styles = useCardStyles()

    return () => (
      <chakra.div __label="card__body" __css={styles.value.body} {...attrs}>
        {slots.default?.()}
      </chakra.div>
    )
  },
})
