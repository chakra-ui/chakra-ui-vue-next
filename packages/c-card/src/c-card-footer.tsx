import {
  type Component,
  type DefineComponent,
  type PropType,
  computed,
  defineComponent,
  h,
} from "vue"
import {
  type HTMLChakraProps,
  chakra,
  type SystemStyleObject,
  type DOMElements,
  ComponentWithProps,
} from "@chakra-ui/vue-system"
import { useCardStyles } from "./card.context"

export interface CCardFooterProps extends HTMLChakraProps<"div"> {
  justify?: SystemStyleObject["justifyContent"]
}

export const CCardFooter = defineComponent({
  name: "CCardFooter",
  props: {
    as: {
      type: [Object, String] as PropType<
        DOMElements | Component | DefineComponent
      >,
      default: "div",
    },
    justify: {
      type: String as PropType<CCardFooterProps["justify"]>,
      default: "",
    },
  },
  setup(props, { attrs, slots }) {
    const styles = useCardStyles()

    const ownStyles = computed(() => ({
      display: "flex",
      justifyContent: props.justify,
      ...styles.value.footer,
    }))

    return () => (
      <chakra.div __label="card__footer" __css={ownStyles.value} {...attrs}>
        {slots.default?.()}
      </chakra.div>
    )
  },
}) as ComponentWithProps<CCardFooterProps>
