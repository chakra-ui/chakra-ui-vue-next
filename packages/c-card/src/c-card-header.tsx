import { defineComponent, h } from "vue"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"
import { useCardStyles } from "./card.context"

export interface CCardHeaderProps extends HTMLChakraProps<"div"> {}

export const CCardHeader = defineComponent({
  name: "CCardHeader",
  setup(_, { attrs, slots }) {
    const styles = useCardStyles()

    return () => (
      <chakra.div __label="card__header" __css={styles.value.header} {...attrs}>
        {slots.default?.()}
      </chakra.div>
    )
  },
})
