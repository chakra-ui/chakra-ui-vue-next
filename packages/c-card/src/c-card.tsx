/**
 * Hey! Welcome to @chakra-ui/vue-next CCard
 *
 * Card is a flexible component used to group and display content in a clear and concise format.
 *
 * @see Docs     https://next.vue.chakra-ui.com/c-card
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/main/packages/c-card/src/c-card.tsx
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import { computed, defineComponent, h, PropType } from "vue"
import {
  chakra,
  DOMElements,
  HTMLChakraProps,
  ThemingProps,
  useMultiStyleConfig,
} from "@chakra-ui/vue-system"
import { vueThemingProps } from "@chakra-ui/vue-utils"
import { CCardStylesProvider } from "./card.context"
import { filterUndefined } from "@chakra-ui/utils"
import { SystemProps } from "@chakra-ui/styled-system"
import type * as CSS from "csstype"

export type CardOptions = {
  /**
   * The flex direction of the card
   */
  direction?: SystemProps["flexDirection"]
  /**
   * The flex alignment of the card
   */
  align?: SystemProps["alignItems"]
  /**
   * The flex distribution of the card
   */
  justify?: SystemProps["justifyContent"]
}

export interface CCardProps
  extends HTMLChakraProps<"div">,
    CardOptions,
    ThemingProps<"Card"> {}

export const CCard = defineComponent({
  name: "CCard",
  props: {
    as: {
      type: [Object, String] as PropType<DOMElements>,
      default: "div",
    },
    direction: {
      type: String as PropType<CCardProps["direction"]>,
      default: "column",
    },
    align: {
      type: String as PropType<CCardProps["align"]>,
    },
    justify: {
      type: String as PropType<CCardProps["justify"]>,
    },
    ...vueThemingProps,
  },
  setup(props, { slots, attrs }) {
    const themingProps = computed(() =>
      filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig,
      })
    )

    const styles = useMultiStyleConfig("Card", themingProps)

    CCardStylesProvider(styles)

    const ownProps = computed(() => ({
      display: "flex",
      flexDirection: props.direction,
      justifyContent: props.justify,
      alignItems: props.align,
      position: "relative",
      minWidth: 0,
      wordWrap: "break-word",
      ...styles.value.container,
    }))

    return () => (
      <chakra.div
        __label="card"
        as={props.as}
        __css={ownProps.value}
        {...attrs}
      >
        {slots.default?.()}
      </chakra.div>
    )
  },
})
