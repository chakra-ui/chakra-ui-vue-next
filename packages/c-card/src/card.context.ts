import { createStylesContext } from "@chakra-ui/vue-system"
import type * as SS from "@chakra-ui/vue-system"

export const [CCardStylesProvider, useCardStyles] = createStylesContext("Card")
