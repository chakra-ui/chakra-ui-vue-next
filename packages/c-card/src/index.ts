export { CCard, type CCardProps } from "./c-card"
export { CCardBody, type CCardBodyProps } from "./c-card-body"
export { CCardHeader, type CCardHeaderProps } from "./c-card-header"
