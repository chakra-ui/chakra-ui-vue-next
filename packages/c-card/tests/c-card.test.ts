import { render } from "../../test-utils/src"
import { CCard, CCardHeader, CCardBody } from "../src"

it("should render properly", () => {
  const { asFragment } = render({
    components: { CCard, CCardHeader, CCardBody },
    template: `
      <c-card>
        <c-card-header>
          Customer dashboard
        </c-card-header>
        <c-card-body>
            View a summary of all your customers over the last month.
        </c-card-body>
      </c-card>
    `,
  })
  expect(asFragment()).toMatchSnapshot()
})
