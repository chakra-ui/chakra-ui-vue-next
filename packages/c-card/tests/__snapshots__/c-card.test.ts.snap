// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <div
    class="chakra-card chakra-1n4p55y"
  >
    <div
      class="chakra-card__header chakra-1sl53ol"
    >
       Customer dashboard 
    </div>
    <div
      class="chakra-card__body chakra-1idwstw"
    >
       View a summary of all your customers over the last month. 
    </div>
  </div>
</DocumentFragment>
`;
